compiler/rustc_borrowck/src/lib.rs

```diff
@@ -235,10 +232,13 @@ fn do_mir_borrowck<'tcx>(
         false
     };
 
-    for (idx, move_data) in promoted_move_data {
+    // While promoteds should mostly be correct by construction, we need to check them for
+    // invalid moves to detect moving out of arrays:`struct S; fn main() { &([S][0]); }`.
+    for promoted_body in &promoted {
         use rustc_middle::mir::visit::Visitor;
-
-        let promoted_body = &promoted[idx];
+        // FIXME: This assumes that we don't try to access most fields of the `promoted_mbcx`.
+        // We may want to move `MoveError` construction and reporting out of `MirBorrowckCtxt`.
+        let move_data = MoveData::gather_moves(body, tcx, |_| true);
```

Comment:
shouldn't this be `promoted_body`?

```suggestion
        let move_data = MoveData::gather_moves(promoted_body, tcx, |_| true);
```

Reply:
If so, then we probably need a test that actually exercises this, because the fact that this didn't fail makes me very nervous lol

Reply:
oh wow, yeah, that's... interesting

Reply:
from what I can tell the answer is: yes, iut should clearly be `promoted_body` and also: it is entirely unused by `check_movable_place` which is very gamer

Reply:
gonna do the refactor from my fixme i guess, because that's sus

```
        // FIXME: This assumes that we don't try to access most fields of the `promoted_mbcx`.
        // We may want to move `MoveError` construction and reporting out of `MirBorrowckCtxt`.
```